### PR TITLE
Clients: support LFNs for declare-bad #42

### PR DESCRIPTION
**Pull request by nikmagini**
Clients: support LFNs for declare-bad #42

### CODE REVIEW

File: bin/rucio-admin [hunk 1]
```diff
@@ -830,7 +831,19 @@ def declare_bad_file_replicas(args):
...
+    for bad_file in bad_files:
+        if bad_file.find('://') == -1:
+            scope, name = extract_scope(bad_file)
+            replicas = [replica for rep in client.list_replicas([{'scope': scope, 'name': name}])
+                        for replica in rep['pfns'].keys()]
+            bad_files_pfns.extend(replicas)
+        else:
+            bad_files_pfns.append(bad_file)
+
+    for chunk in chunks(bad_files_pfns, 500):
```
Comment (cserf) [inline bin/rucio-admin]: There might be a problem in case the PFNs have different schemes

Comment (nikmagini) [reply bin/rucio-admin]: OK - should I force one scheme in particular e.g. 'srm'?
Or I could skip all replicas that don't have the same scheme as the first item in the list.

File: bin/rucio-admin [hunk 2]
```diff
@@ -1192,9 +1205,9 @@ Commands:
     # The declare-bad command
     declare_bad_file_replicas_parser = rep_subparser.add_parser('declare-bad', help='Declare bad file replicas')
     declare_bad_file_replicas_parser.set_defaults(which='declare_bad_file_replicas')
-    declare_bad_file_replicas_parser.add_argument(dest='listbadfiles', action='store', nargs='*', help='The list of bad files')
+    declare_bad_file_replicas_parser.add_argument(dest='listbadfiles', action='store', nargs='*', help='List of bad file PFNs (for one replica) or LFNs (for all replicas)')
     declare_bad_file_replicas_parser.add_argument('--reason', dest='reason', required=True, action='store', help='Reason')
-    declare_bad_file_replicas_parser.add_argument('--inputfile', dest='inputfile', nargs='?', action='store', help='Reason')
+    declare_bad_file_replicas_parser.add_argument('--inputfile', dest='inputfile', nargs='?', action='store', help='File containing list of bad files')
```
Comment (mlassnig) [inline bin/rucio-admin]: from the discussion today, maybe something like "--allow-collection" to enable marking of datasets and containers

Comment (mlassnig) [reply bin/rucio-admin]: @cserf opinion?

Comment (bari12) [reply bin/rucio-admin]: Yes. I think a simple check if the input did is a FILE or DATASET/CONTAINER. If dataset/container the --allow-collection has to be used.

File: bin/rucio-admin (hunk 1 above)
Comment (nikmagini) [reply bin/rucio-admin]: Now the list of PFNs is split into separate sublists - one sublist for each scheme

File: bin/rucio-admin (hunk 2 above)
Comment (nikmagini) [reply bin/rucio-admin]: Implemented --allow-collection safeguard.
One last point: maybe the list of PFNs that will be declared bad should be printed out? Particularly if it's generated from a DID

Comment (bari12) [reply bin/rucio-admin]: I think a print out would be good yes.

Comment (nikmagini) [reply bin/rucio-admin]: done!

**Comment by bari12** [comment]
I think you might have to git rebase your branch against master and push again to your branch to kick travis to start the test.

**Comment by nikmagini** [comment]
Done, it seems that Travis is running now

File: bin/rucio-admin [hunk 3]
```diff
@@ -839,11 +841,35 @@ def declare_bad_file_replicas(args):
...
+                return FAILURE
+            replicas = [replica for rep in client.list_replicas([{'scope': scope, 'name': name}])
+                        for replica in rep['pfns'].keys()]
+            bad_files_pfns.extend(replicas)
+        else:
+            bad_files_pfns.append(bad_file)
+    print "PFNs that will be declared bad:"
+    for pfn in bad_files_pfns:
+        print pfn
+
```
Comment (cserf) [inline bin/rucio-admin]: Maybe you should remove the print statement there, or at least make them optional

**Review by cserf — COMMENTED** [review]
Just one comment about print statement. The rest is OK

File: bin/rucio-admin (hunk 3 above)
Comment (nikmagini) [reply bin/rucio-admin]: Ok, suppressed when not using --verbose

**Comment by nikmagini** [comment]
It seems that I got travis stuck again when I pushed updates...

**Comment by bari12** [comment]
Yes, you might have to rebase again. I merged some other PRs in the meanwhile.

**Comment by vingar** [comment]
Could you try to commit again with the --amend option or squash several commits in one and force push the branch ?

**Comment by nikmagini** [comment]
@vingar I did squash everything into a single commit and force-push already

**Comment by bari12** [comment]
I think this might be because of the conflict in the other PR against next. (Nothing you can do against this) - Somehow travis handles them internally together. I will test this manually.

**Comment by bari12** [comment]
Tests run through fine. I am force merging this.

**Pull request merged**